iceoryx_utils/source/relocatable_pointer/base_relocatable_ptr.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/internal/relocatable_pointer/base_relocatable_ptr.hpp"
+
+namespace iox
+{
+namespace rp
+{
+BaseRelocatablePointer::BaseRelocatablePointer() noexcept
+{
+}
+
+BaseRelocatablePointer::BaseRelocatablePointer(const void* ptr) noexcept
+    : m_offset(computeOffset(ptr))
+{
+}
+
+BaseRelocatablePointer::BaseRelocatablePointer(const BaseRelocatablePointer& other) noexcept
+    : m_offset(computeOffset(other.computeRawPtr()))
+{
+}
+
```

Comment:
virtual d'tor for BaseRelocatablePointer + for every base class

Reply:
nope, this lives in the shared memory, therefore virtual is not allowed

Reply:
Virtual dtor also only if you use virtual functions to ensure the correct dtor is called. This is not an issue here since we do not use them.

Reply:
Why do we inherit at all here? Wouldn't it be possible to use it as a composite then we would be completely on the safe side.
I agree with you but this can be easily error prone - not without reason Scott Meyers suggest to always have a virtual destructor when working with inheritance.

Reply:
Can we maybe also rename the `BaseRelocatablePointer` to `UntypedRelocatablePointer` or `RelocatablePointerUntyped`?

Ideally this would be the same type and just an internal member would be specialized for void.

Reply:
Using composition instead of inheritance should be part of a refactoring of the relocatable pointer. Therefore I added this point to the issue. I'd prefer to do the renaming in this context as well.

Reply:
I had another idea. Will continue the discussion in the issue.

Reply:
The inheritance can be avoided, that is correct. Historically the reason is hat I started with an untyped version (as @elBoberido observed), which is currently the base. For the sake of code reuse I inherited in the typed version but I would now solve it entirely with templates, neither composite nor inheritance I think. This is a larger refactoring which I could start after 1.0, depending on priorities.
But functionality-wise it is no priority since I am not aware of bugs. I will improve the generality and the design in the future.

Reply:
@MatthiasKillat I made a proposal in the issue itself